### PR TITLE
Add notification use case

### PR DESCRIPTION
**Pull request by 9a24f0**
The use case is very draft as I could not imagine what will happens and from whom Student will receive notifications from.
Any helps is highly appreciated!

Closing GH-21 when this is merged.

### CODE REVIEW

**Comment by Huy-Ngo** [comment]
> from whom Student will receive notifications from

From the system, I think. Like when someone comment on zer post, ze will receive a notification, for example.

File: docs/source/requirements/use-case-model.rst [hunk 1]
```diff
@@ -511,6 +513,52 @@ Extension Points
...
+Special Requirements
+^^^^^^^^^^^^^^^^^^^^
+
+Student could either receive notification from Lecturer or
+subscribed disscussion threads.
+
+Pre-Conditions
+^^^^^^^^^^^^^^
+
+User must be Student.
```
Comment (McSinyx) [inline docs/source/requirements/use-case-model.rst:548]: The use-case already specifies the actor.  Conditions refer to states, e.g. is the actor logged in (i.e. is the system storing a session of the actor) or something.

File: docs/source/requirements/use-case-model.rst [hunk 2]
```diff
@@ -511,6 +513,52 @@ Extension Points
...
+
+Post-Conditions
+^^^^^^^^^^^^^^^
+
+Student gets an unread notification.
```
Comment (McSinyx) [inline docs/source/requirements/use-case-model.rst:553]: Same here, we care about the system state.  Student getting notif is part of the flow.

File: docs/source/solution/database.rst [hunk 3]
```diff
@@ -26,8 +26,8 @@ The pattern for user objects is described below:
       "email": "foo.bar@example.com",
       "password-hash": "(hash generated for password)",
       "role": "student",
-      "major": "ICT", // optional
-      "bio": "A self description" // optional
+      "major": "ICT (optional)",
+      "bio": "A self description (optional)"
```
Comment (McSinyx) [inline docs/source/solution/database.rst:30]: Noice.

File: docs/source/requirements/use-case-model.rst [hunk 4]
```diff
@@ -511,6 +513,52 @@ Extension Points
...
+
+Extension Points
+^^^^^^^^^^^^^^^^
+
+None.
+
+
+
```
Comment (McSinyx) [inline docs/source/requirements/use-case-model.rst:561]:
```suggestion
```

**Review by McSinyx — CHANGES_REQUESTED** [review]

File: docs/source/requirements/use-case-model.rst (hunk 2 above)
Comment (9a24f0) [reply docs/source/requirements/use-case-model.rst:553]: What I mean here is the system change, take notification of gh for example: when we receive a notification, the state changes to: has unread notif. After that, when we reads to notif, the state changes to no unread notif.

Maybe what I wrote there was misleading :smile:

Comment (McSinyx) [reply docs/source/requirements/use-case-model.rst:553]: Maybe rephrase it to passive tense of the notification?

File: docs/source/requirements/use-case-model.rst [hunk 5]
```diff
@@ -511,6 +513,52 @@ Extension Points
...
+This use case allows Student to receive notification from
+both in-app system and email.
+
+Flow of Events
+^^^^^^^^^^^^^^
+
+Basic Flow
+""""""""""
+
+1. System sends notification to Student.
```
[the remark this answers is not in the document]
Comment (Huy-Ngo) [reply docs/source/requirements/use-case-model.rst:531]: Yea, I agree. Some that come up in my mind are:

- A person ask to join your project
- Join request is approved
- The deadline is near
- Assigned to a task
- Task is approved/evaluated
- Project is evaluated
- Reply in discussion

[the remark this answers is not in the document]
Comment (9a24f0) [reply docs/source/requirements/use-case-model.rst:531]: IMHO, we could not include "the trigger" in basic flow simply because it is not basic at all. I'm thinking of putting it in pre-conditions but then I chose special requirements.

File: docs/source/solution/database.rst (hunk 3 above)
Comment (Huy-Ngo) [reply docs/source/solution/database.rst:30]: @9a24f0 Please undo these changes. I'm going to remove these anyway so that may lead to conflicts.

Comment (9a24f0) [reply docs/source/solution/database.rst:30]: feel free to drop a change request, don't comment on something that is mark resolved bc it took me 1 hour to find your comment :sweat:

File: docs/source/requirements/use-case-model.rst [hunk 6]
```diff
@@ -511,6 +513,58 @@ Extension Points
...
+- The deadline for his/her tasks is near.
+- He/she is assigned to a task.
+- His/her task is approved/evaluated.
+- His/her project is evaluated
+- There is a new reply in his/her watching discussion thread.
+
+Post-Conditions
+^^^^^^^^^^^^^^^
+
+A new unread notification is received.
```
Comment (McSinyx) [inline docs/source/requirements/use-case-model.rst:560]: I think you may want to main it is *marked as read* or *removed*, depending on how you want it to be.  FWIW with emails this is quite difficult/awkward to accomplished.

Comment (9a24f0) [reply docs/source/requirements/use-case-model.rst:560]: erm... could you rephrase that

**Review by McSinyx — APPROVED** [review]

File: docs/source/requirements/use-case-model.rst (hunk 6 above)
Comment (McSinyx) [reply docs/source/requirements/use-case-model.rst:560]: It's all good, no worries, my brain is not functioning as well as it should :smile:

**Review by DDHLong — APPROVED** [review]
